False positives using when using pytest-xdist -f (--looponfail)

When doing TDD, after the first mypy error, the plugin start emitting errors on collection.

It will continue normally until you hit an actual mypy fail, then in the follow round (after mypy error is fixed) this warning is emitted forever until pytest is killed and restarted.

Steps to Reproduce (assumes python 3.8 and pipenv available):
1. `mkdir ~/tmp20200412`
2. `cd ~/tmp20200412`
3. `git clone git@bitbucket.org:y2kbugger/test_driven_development_by_example_tdd_follow_along.git`
4. `cd ./test*`
5. `git checkout b0a2f56`
6. replicate versions exactly from Pipfile.lock `pipenv sync`
7. put venv on PATH `pipenv shell`
8. run pytest `pytest --mypy -f money.py test_money.py`

In another terminal:
1. `cd ~/tmp20200412/test*`
2. cause mypy error `echo 'lol: str = 2' >> money.py`
3. undo mypy error `git checkout b0a2f56 -- .`

```
###################################### waiting for changes #######################################
### Watching:   /home/y2k/devel/test_driven_development_by_example_tdd_follow_along
# MODIFIED /home/y2k/devel/test_driven_development_by_example_tdd_follow_along/money.py
============================= test session starts ==============================
platform linux -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/y2k/devel/test_driven_development_by_example_tdd_follow_along
plugins: forked-1.1.3, mypy-0.6.1, xdist-1.31.0, flake8-1.0.4
collected 0 items / 1 error
collected 0 items / 1 error

==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
../../.local/share/virtualenvs/test_driven_development_by_example_tdd_fol-TPMT2dI5/lib/python3.8/site-packages/_pytest/runner.py:244: in from_call
    result = func()
../../.local/share/virtualenvs/test_driven_development_by_example_tdd_fol-TPMT2dI5/lib/python3.8/site-packages/_pytest/runner.py:264: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
../../.local/share/virtualenvs/test_driven_development_by_example_tdd_fol-TPMT2dI5/lib/python3.8/site-packages/_pytest/main.py:498: in collect
    yield from self._collect(fspath, parts)
../../.local/share/virtualenvs/test_driven_development_by_example_tdd_fol-TPMT2dI5/lib/python3.8/site-packages/_pytest/main.py:537: in _collect
    assert not names, "invalid arg {!r}".format((argpath, names))
E   AssertionError: invalid arg (local('/home/y2k/devel/test_driven_development_by_example_tdd_follow_along'), ['mypy'])
=========================== short test summary info ============================
ERROR  - AssertionError: invalid arg (local('/home/y2k/devel/test_driven_deve...
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.22s ===============================
######################################### LOOPONFAILING ##########################################
money.py::mypy
test_money.py::test_multiplication
test_money.py::test_franc_multiplication
test_money.py::test_equality
test_money.py::test_currency
::mypy
###################################### waiting for changes #######################################
### Watching:   /home/y2k/devel/test_driven_development_by_example_tdd_follow_along
```
Note I kill with keyboard interrupt here:
```
^C
```
Then everything goes fine again:

```
(test_driven_development_by_example_tdd_follow_along) [y2k@15r test_driven_development_by_example_tdd_follow_along]
$ pytest --mypy -f money.py test_money.py 
============================= test session starts ==============================
platform linux -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/y2k/devel/test_driven_development_by_example_tdd_follow_along
plugins: forked-1.1.3, mypy-0.6.1, xdist-1.31.0, flake8-1.0.4
collected 7 items
collected 7 items

money.py .                                                               [ 12%]
test_money.py ......                                                     [ 87%]
. .                                                                      [100%]. [100%]
===================================== mypy =====================================

Success: no issues found in 2 source files
============================== 9 passed in 0.34s ===============================
###################################### waiting for changes #######################################
### Watching:   /home/y2k/devel/test_driven_development_by_example_tdd_follow_along
```

When a test fails, `pytest-xdist` holds onto the node ID of the failing test. The `MypyStatusItem` that is [injected in `pytest_collection_modifyitems`](https://github.com/dbader/pytest-mypy/blob/573667535d1f6f731d76fd87f1360b9de85f7aa7/src/pytest_mypy.py#L113-L130) has `parent=session` which produces a node ID of `::mypy` (where `mypy` is `nodeid_name`).
Running that node ID explicitly (like `xdist` does) reproduces this:
``` sh
pytest --verbose --mypy ::mypy
```
```
================================================ test session starts =================================================
platform linux -- Python 3.8.2, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- /home/dtux/Projects/pytest-mypy/venv/bin/python
cachedir: .pytest_cache
Using --randomly-seed=1597611273
rootdir: /home/dtux/Projects/pytest-mypy
plugins: xdist-2.0.0, forked-1.3.0, randomly-2.1.1, cov-2.5.1, mypy-0.6.3.dev21+gd052c4c.d20200816
collected 0 items / 1 error                                                                                          

======================================================= ERRORS =======================================================
___________________________________________ ERROR collecting test session ____________________________________________
venv/lib/python3.8/site-packages/_pytest/runner.py:294: in from_call
    result = func()  # type: Optional[_T]
venv/lib/python3.8/site-packages/_pytest/runner.py:324: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
venv/lib/python3.8/site-packages/_pytest/main.py:576: in collect
    yield from self._collect(fspath, parts)
venv/lib/python3.8/site-packages/_pytest/main.py:617: in _collect
    assert not names, "invalid arg {!r}".format((argpath, names))
E   AssertionError: invalid arg (local('/home/dtux/Projects/pytest-mypy'), ['mypy'])
============================================== short test summary info ===============================================
ERROR  - AssertionError: invalid arg (local('/home/dtux/Projects/pytest-mypy'), ['mypy'])
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================== 1 error in 0.08s ==================================================
```

--- a/src/pytest_mypy.py
+++ b/src/pytest_mypy.py
@@ -125,26 +125,16 @@ class MypyFile(pytest.File):
     def collect(self):
         """Create a MypyFileItem for the File."""
         yield MypyFileItem.from_parent(parent=self, name=nodeid_name)
-
-
-@pytest.hookimpl(hookwrapper=True, trylast=True)
-def pytest_collection_modifyitems(session, config, items):
-    """
-    Add a MypyStatusItem if any MypyFileItems were collected.
-
-    Since mypy might check files that were not collected,
-    pytest could pass even though mypy failed!
-    To prevent that, add an explicit check for the mypy exit status.
-
-    This should execute as late as possible to avoid missing any
-    MypyFileItems injected by other pytest_collection_modifyitems
-    implementations.
-    """
-    yield
-    if any(isinstance(item, MypyFileItem) for item in items):
-        items.append(
-            MypyStatusItem.from_parent(parent=session, name=nodeid_name),
-        )
+        # Since mypy might check files that were not collected,
+        # pytest could pass even though mypy failed!
+        # To prevent that, add an explicit check for the mypy exit status.
+        if not any(
+                isinstance(item, MypyStatusItem) for item in self.session.items
+        ):
+            yield MypyStatusItem.from_parent(
+                parent=self,
+                name=nodeid_name + "-status",
+            )
 
 
 class MypyItem(pytest.Item):

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -214,38 +214,6 @@ def test_api_nodeid_name(testdir, xdist_args):
     assert result.ret == 0
 
 
-def test_pytest_collection_modifyitems(testdir, xdist_args):
-    """
-    Verify that collected files which are removed in a
-    pytest_collection_modifyitems implementation are not
-    checked by mypy.
-
-    This would also fail if a MypyStatusItem were injected
-    despite there being no MypyFileItems.
-    """
-    testdir.makepyfile(conftest='''
-        def pytest_collection_modifyitems(session, config, items):
-            plugin = config.pluginmanager.getplugin('mypy')
-            for mypy_item_i in reversed([
-                    i
-                    for i, item in enumerate(items)
-                    if isinstance(item, plugin.MypyFileItem)
-            ]):
-                items.pop(mypy_item_i)
-    ''')
-    testdir.makepyfile('''
-        def pyfunc(x: int) -> str:
-            return x * 2
-
-        def test_pass():
-            pass
-    ''')
-    result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    test_count = 1
-    result.assert_outcomes(passed=test_count)
-    assert result.ret == 0
-
-
 def test_mypy_indirect(testdir, xdist_args):
     """Verify that uncollected files checked by mypy cause a failure."""
     testdir.makepyfile(bad='''
@@ -256,38 +224,6 @@ def test_mypy_indirect(testdir, xdist_args):
         import bad
     ''')
     result = testdir.runpytest_subprocess('--mypy', *xdist_args, 'good.py')
-    assert result.ret != 0
-
-
-def test_mypy_indirect_inject(testdir, xdist_args):
-    """
-    Verify that uncollected files checked by mypy because of a MypyFileItem
-    injected in pytest_collection_modifyitems cause a failure.
-    """
-    testdir.makepyfile(bad='''
-        def pyfunc(x: int) -> str:
-            return x * 2
-    ''')
-    testdir.makepyfile(good='''
-        import bad
-    ''')
-    testdir.makepyfile(conftest='''
-        import py
-        import pytest
-
-        @pytest.hookimpl(trylast=True)  # Inject as late as possible.
-        def pytest_collection_modifyitems(session, config, items):
-            plugin = config.pluginmanager.getplugin('mypy')
-            items.append(
-                plugin.MypyFileItem.from_parent(
-                    parent=session,
-                    name=str(py.path.local('good.py')),
-                ),
-            )
-    ''')
-    name = 'empty'
-    testdir.mkdir(name)
-    result = testdir.runpytest_subprocess('--mypy', *xdist_args, name)
     assert result.ret != 0
 
 

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -1,3 +1,6 @@
+import signal
+import textwrap
+
 import pytest
 
 
@@ -269,3 +272,87 @@ def test_setup_cfg(testdir, xdist_args):
         '1: error: Function is missing a type annotation',
     ])
     assert result.ret != 0
+
+
+def test_looponfail(testdir):
+    """Ensure that the plugin works with --looponfail."""
+
+    pass_source = textwrap.dedent(
+        """\
+        def pyfunc(x: int) -> int:
+            return x * 2
+        """,
+    )
+    fail_source = textwrap.dedent(
+        """\
+        def pyfunc(x: int) -> str:
+            return x * 2
+        """,
+    )
+    pyfile = testdir.makepyfile(fail_source)
+    looponfailroot = testdir.mkdir("looponfailroot")
+    looponfailroot_pyfile = looponfailroot.join(pyfile.basename)
+    pyfile.move(looponfailroot_pyfile)
+    pyfile = looponfailroot_pyfile
+    testdir.makeini(
+        textwrap.dedent(
+            """\
+            [pytest]
+            looponfailroots = {looponfailroots}
+            """.format(
+                looponfailroots=looponfailroot,
+            ),
+        ),
+    )
+
+    child = testdir.spawn_pytest(
+        "--mypy --looponfail " + str(pyfile),
+        expect_timeout=30.0,
+    )
+
+    def _expect_session():
+        child.expect("==== test session starts ====")
+
+    def _expect_failure():
+        _expect_session()
+        child.expect("==== FAILURES ====")
+        child.expect(pyfile.basename + " ____")
+        child.expect("2: error: Incompatible return value")
+        # These only show with mypy>=0.730:
+        # child.expect("==== mypy ====")
+        # child.expect("Found 1 error in 1 file (checked 1 source file)")
+        child.expect("2 failed")
+        child.expect("#### LOOPONFAILING ####")
+        _expect_waiting()
+
+    def _expect_waiting():
+        child.expect("#### waiting for changes ####")
+        child.expect("Watching")
+
+    def _fix():
+        pyfile.write(pass_source)
+        _expect_changed()
+        _expect_success()
+
+    def _expect_changed():
+        child.expect("MODIFIED " + str(pyfile))
+
+    def _expect_success():
+        for _ in range(2):
+            _expect_session()
+            # These only show with mypy>=0.730:
+            # child.expect("==== mypy ====")
+            # child.expect("Success: no issues found in 1 source file")
+            child.expect("2 passed")
+        _expect_waiting()
+
+    def _break():
+        pyfile.write(fail_source)
+        _expect_changed()
+        _expect_failure()
+
+    _expect_failure()
+    _fix()
+    _break()
+    _fix()
+    child.kill(signal.SIGTERM)

--- a/tox.ini
+++ b/tox.ini
@@ -132,6 +132,8 @@ deps =
     mypy0.79: mypy >= 0.790, < 0.800
     mypy0.7x: mypy >= 0.700, < 0.800
 
+    pexpect ~= 4.8.0
+
 commands = py.test -p no:mypy --cov pytest_mypy --cov-fail-under 100 --cov-report term-missing {posargs:-n auto} tests
 
 [testenv:publish]

